libs/wingc/src/inflight_transform.rs

```diff
@@ -0,0 +1,277 @@
+							statements: Scope::new(vec![], expr.span.clone()),
+							span: expr.span.clone(),
+						},
+						fields: resource_fields,
+						implements: vec![],
+						parent: None,
+						methods: vec![(handle_name.clone(), new_func_def)],
+						inflight_initializer: None,
+					}),
+					idx: 1,
+					span: expr.span.clone(),
+				};
+
+				// return_resource_instance :=
+				// ```
+				// return new <resource_name>();
+				// ```
+				let return_resource_instance = Stmt {
+					kind: StmtKind::Return(Some(Expr::new(
+						ExprKind::New {
+							class: resource_type_annotation,
+							arg_list: ArgList {
+								named_args: IndexMap::new(),
+								pos_args: vec![],
+							},
+							obj_id: None,
+							obj_scope: None,
+						},
+						expr.span.clone(),
+					))),
+					idx: 2,
+					span: expr.span.clone(),
+				};
+
+				// make_resource_body :=
+				// ```
+				// {
+				//   <parent_this_def>? // only if we are inside a resource
+				//   <resource_def>
+				//   <return_resource_instance>
+				// }
+				// ```
+				let make_resource_body = if self.is_inside_resource {
+					// parent_this_def :=
+					// ```
+					// let __parent_this = this;
+					// ```
+					let parent_this_def = Stmt {
+						kind: StmtKind::VariableDef {
+							reassignable: false,
+							var_name: parent_this_name,
+							initial_value: Expr::new(ExprKind::Reference(Reference::Identifier(this_name)), expr.span.clone()),
+							type_: None, // thank god we have type inference
+						},
+						idx: 0,
+						span: expr.span.clone(),
+					};
+
+					Scope::new(
+						vec![parent_this_def, resource_def, return_resource_instance],
+						expr.span.clone(),
+					)
+				} else {
+					Scope::new(vec![resource_def, return_resource_instance], expr.span.clone())
+				};
+
+				// make_resource_closure :=
+				// ```
+				// (): resource => { ...<make_resource_body> }
+				// ```
+				let make_resource_closure = Expr::new(
+					ExprKind::FunctionClosure(FunctionDefinition {
+						signature: FunctionSignature {
+							parameters: vec![],
+							return_type: Some(Box::new(TypeAnnotation {
+								kind: TypeAnnotationKind::Resource,
+								span: expr.span.clone(),
+							})),
+							phase: Phase::Preflight,
+						},
+						body: crate::ast::FunctionBody::Statements(make_resource_body),
+						is_static: false,
+						span: expr.span.clone(),
+					}),
+					expr.span.clone(),
+				);
+
+				// call_expr :=
+				// ```
+				// <make_resource_closure>()
+				// ```
+				let call_expr = Expr::new(
+					ExprKind::Call {
+						function: Box::new(make_resource_closure),
+						arg_list: ArgList::new(),
+					},
+					expr.span,
+				);
+				call_expr
+			}
+			_ => fold::fold_expr(self, expr),
+		}
+	}
+}
+
+/// Transform usages of "this.<field>" into "__parent_this.<field>"
+struct ThisTransformer;
+
+impl Fold for ThisTransformer {
+	fn fold_reference(&mut self, node: Reference) -> Reference {
+		match node {
+			Reference::Identifier(ident) => {
```

Comment:
See my other comment about making it possible to code generate the `Fold` boilerplate (future work)

I think the clean way to achieve this would be to have the `Reference::Identifier` enum contain `Identifier` (a new struct that wraps a string) instead of `String`. like so:

```rs
pub enum Reference {
	Identifier(Identifier),
	InstanceMember(InstanceMember),
	TypeMember(TypeMember),
}

pub struct Identifier(String);
pub struct InstanceMember(/* snip */);
```